File: 0002-add-two-numbers/0002-add-two-numbers.java
class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        ListNode dummy = new ListNode();
        ListNode start = dummy;
        
        int carry = 0;
        while (l1 != null || l2 != null || carry != 0) {
            int res = l1 == null ? 0 : l1.val;
            res += l2 == null ? 0 : l2.val;
            res += carry;
            start.next = new ListNode(res % 10);
            start = start.next;
            carry = res / 10;
            if (l1 != null) l1 = l1.next;
            if (l2 != null) l2 = l2.next;
        }
        
        return dummy.next;
    }
}

File: 0002-add-two-numbers/0002-add-two-numbers.cpp
class Solution {
public:
    ListNode* reverseLL(ListNode* cur) {
        ListNode* prev = nullptr;
        
        while (cur != nullptr) {
            ListNode* front = cur->next;
            cur->next = prev;
            prev = cur;
            cur = front;
        }
        
        return prev;
    }
    
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        ListNode* res = new ListNode(0);
        ListNode* p = res;
        
        int carry = 0;
        
        while (l1 != nullptr || l2 != nullptr || carry != 0) {
            int sum = carry;
            
            if (l1 != nullptr) {
                sum = sum + l1->val;
                l1 = l1->next;
            }
            
            if (l2 != nullptr) {
                sum = sum + l2->val;
                l2 = l2->next;
            }
            
            p->next = new ListNode(sum % 10);
            carry = sum / 10;
            
            p = p->next;
        }
        
        return res->next;
    }
};